inc/Engine/Optimization/RUCSS/Warmup/ResourceFetcher.php

```diff
@@ -0,0 +1,175 @@
+<?php
+declare(strict_types=1);
+
+namespace WP_Rocket\Engine\Optimization\RUCSS\Warmup;
+
+use WP_Rocket\Engine\Optimization\AbstractOptimization;
+use WP_Rocket\Engine\Optimization\AssetsLocalCache;
+use WP_Rocket\Logger\Logger;
+
+class ResourceFetcher extends AbstractOptimization {
+
+	/**
+	 * Resources array.
+	 *
+	 * @var array
+	 */
+	private $resources = [];
+
+	/**
+	 * Assets local cache instance
+	 *
+	 * @var AssetsLocalCache
+	 */
+	private $local_cache;
+
+	/**
+	 * Resource fetcher process instance.
+	 *
+	 * @var ResourceFetcherProcess
+	 */
+	private $process;
+
+	/**
+	 * Resource constructor.
+	 *
+	 * @param AssetsLocalCache       $local_cache Local cache instance.
+	 * @param ResourceFetcherProcess $process     Resource fetcher process instance.
+	 */
+	public function __construct( AssetsLocalCache $local_cache, ResourceFetcherProcess $process ) {
+		$this->local_cache = $local_cache;
+		$this->process     = $process;
+	}
+
+	/**
+	 * Handle Collecting resources and save them into the DB
+	 *
+	 * @param string $html Page HTML.
+	 */
+	public function handle( $html ) {
+		// Grab resources from page HTML.
+		$this->find_styles( $html );
+		$this->find_scripts( $html );
+
+		if ( empty( $this->resources ) ) {
+			return;
+		}
+
+		// Send resources to the background process to be saved into DB.
+		foreach ( $this->resources as $resource ) {
+			$this->process->push_to_queue( $resource );
+		}
+
+		$this->process->save()->dispatch();
+	}
+
+	/**
+	 * Find link styles on the page HTML.
+	 *
+	 * @param string $html Page HTML.
+	 */
+	private function find_styles( $html ) {
+		$links = $this->find( '<link(?:[^>]+[\s"\'])?href\s*=\s*[\'"]\s*(?<url>[^\'"\s]+)\s*?[\'"](?:[^>]+)?\/?>', $html );
+
+		if ( empty( $links ) ) {
+			return;
+		}
+
+		foreach ( $links as $link ) {
+			if ( ! (bool) preg_match( '/rel=[\'"]stylesheet[\'"]/is', $link[0] ) ) {
```

Comment:
Is there a reason to bailout in case there is no `rel=stylesheet` on the link?

Reply:
In the first regex above, we've collected ALL the link elements (theoretically). Now we only want the ones that are stylesheets here (as opposed to, say, links for prefetches, etc.). That said, it may be more efficient to eliminate them in the first pass with the previous regex rather than re-loop and check them here.

Reply:
Ah, you are right @iCaspar. But I don't think we have an easy way to identify it in the REGEX due to all potential order of attributes. I need to update the code in Phase 3 too :)

Reply:
Indeed, the regex to do it in 1 go would be a big hairy monster 👹

Reply:
so this is resolved :D